RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc

```diff
@@ -94,4 +107,151 @@ void PatternRecognitionbyCA::makeTracksters(const PatternRecognitionAlgoBase::In
                                     << " count: " << (int)trackster.vertex_multiplicity[i] << std::endl;
     }
   }
+
+  // energy regression and ID when a session is created
+  if (eidSession_ != nullptr) {
+    energyRegressionAndID(input.layerClusters, result);
+  }
+}
+
+void PatternRecognitionbyCA::energyRegressionAndID(const std::vector<reco::CaloCluster> &layerClusters,
+                                                   std::vector<Trackster> &tracksters) {
+  // Energy regression and particle identification strategy:
+  //
+  // 1. Set default values for regressed energy and particle id for each trackster.
+  // 2. Store indices of tracksters whose total sum of cluster energies is above the
+  //    eidMinClusterEnergy_ (GeV) treshold. Inference is not applied for soft tracksters.
+  // 3. When no trackster passes the selection, return.
+  // 4. Create input and output tensors. The batch dimension is determined by the number of
+  //    selected tracksters.
+  // 5. Fill input tensors with layer cluster features. Per layer, clusters are ordered descending
+  //    by energy. Given that tensor data is contiguous in memory, we can use pointer arithmetic to
+  //    fill values, even with batching.
+  // 6. Zero-fill features for empty clusters in each layer.
+  // 7. Batched inference.
+  // 8. Assign the regressed energy and id probabilities to each trackster.
+  //
+  // Indices used throughout this method:
+  // i -> batch / trackster
+  // j -> layer
+  // k -> cluster
+  // l -> feature
+
+  // set default values per trackster, determine if the cluster energy threshold is passed,
+  // and store indices of hard tracksters
+  std::vector<int> tracksterIndices;
+  for (int i = 0; i < (int)tracksters.size(); i++) {
+    // set default values (1)
```

Comment:
it looks like range loops should work well here and in the other loops in the scope

Reply:
1st comment: You're right. The entire NN inference part was intended to be optional, but I think we should make it mandatory and throw an exception at this point.

2nd comment: I actually need the index a few lines below to store the positions of selected tracksters. But I will change the loops in the scope where possible.